### PR TITLE
MLRun CE Deployer Script

### PR DESCRIPTION
**Pull request by quaark**
CLI tool for deploying MLRun Community Edition.
The CLI supports 3 commands:
- `deploy`: Deploys (or upgrades) an MLRun Community Edition Stack.
- `delete`: Uninstalls the CE and cleans up dangling resources.
- `patch-minikube-images`: If using custom images and running from Minikube, this command will patch the images to the Minikube env.

Command Usage:
```
$ python automation/deployment/ce.py --help       
Usage: ce.py [OPTIONS] COMMAND [ARGS]...

  MLRun Community Edition Deployment CLI Tool

Options:
  --help  Show this message and exit.

Commands:
  delete                 Uninstall MLRun Community Edition Deployment
  deploy                 Deploy MLRun Community Edition
  patch-minikube-images  Patch MLRun Community Edition Deployment images...
```

Deploy:
```
$ python automation/deployment/ce.py deploy --help               
Usage: ce.py deploy [OPTIONS]

  Deploy (or upgrade) MLRun Community Edition

Options:
  -cv, --chart-version TEXT       Version of the mlrun chart to install. If
                                  not specified, will install the latest
                                  version
  --devel                         Get the latest RC version of the mlrun
                                  chart. (Only works if --chart-version is not
                                  specified)
  --disable-pipelines             Disable the installation of Kubeflow
                                  Pipelines
  --disable-prometheus-stack      Disable the installation of the Prometheus
                                  stack
  --disable-spark-operator        Disable the installation of the Spark
                                  operator
  -f, --log-file TEXT             Path to log file. If not specified, will log
                                  only to stdout
  -m, --minikube                  Install the mlrun chart in local minikube.
  -mv, --mlrun-version TEXT       Version of mlrun to install. If not
                                  specified, will install the latest version
  -n, --namespace TEXT            Namespace to install the platform in.
                                  Defaults to 'mlrun'
  --override-jupyter-image TEXT   Override the jupyter image. Format:
                                  <repo>:<tag>
  --override-mlrun-api-image TEXT
                                  Override the mlrun-api image. Format:
                                  <repo>:<tag>
  --override-mlrun-ui-image TEXT  Override the mlrun-ui image. Format:
                                  <repo>:<tag>
  --registry-password TEXT        Password of the container registry to use
                                  for storing images
  --registry-secret-name TEXT     Name of the secret containing the
                                  credentials for the container registry to
                                  use for storing images
  --registry-url TEXT             URL of the container registry to use for
                                  storing images  [required]
  --registry-username TEXT        Username of the container registry to use
                                  for storing images
  --set TEXT                      Set custom values for the mlrun chart.
                                  Format: <key>=<value>
  --sqlite TEXT                   Path to sqlite file to use as the mlrun
                                  database. If not supplied, will use MySQL
                                  deployment
  --upgrade                       Upgrade the existing mlrun installation.
  -v, --verbose                   Enable debug logging
  --help                          Show this message and exit.
```

Delete:
```
$ python automation/deployment/ce.py delete --help
Usage: ce.py delete [OPTIONS]

  Uninstall MLRun Community Edition Deployment

Options:
  --cleanup-namespace             Delete the namespace created during
                                  installation. This overrides the other
                                  cleanup options. WARNING: This will result
                                  in data loss!
  --cleanup-volumes               Delete the PVCs created during installation.
                                  WARNING: This will result in data loss!
  -f, --log-file TEXT             Path to log file. If not specified, will log
                                  only to stdout
  -n, --namespace TEXT            Namespace to install the platform in.
                                  Defaults to 'mlrun'
  --registry-secret-name TEXT     Name of the secret containing the
                                  credentials for the container registry to
                                  use for storing images
  --skip-cleanup-registry-secret  Skip deleting the registry secret created
                                  during installation
  --skip-uninstall                Skip uninstalling the Helm chart. Useful if
                                  already uninstalled and you want to perform
                                  cleanup only
  --sqlite TEXT                   Path to sqlite file to use as the mlrun
                                  database. If not supplied, will use MySQL
                                  deployment
  -v, --verbose                   Enable debug logging
  --help                          Show this message and exit.
```

Patch Minikube Images:
```
$ python automation/deployment/ce.py patch-minikube-images --help
Usage: ce.py patch-minikube-images [OPTIONS]

  Patch MLRun Community Edition Deployment images to minikube. Useful if
  overriding images and running in minikube

Options:
  --jupyter-image TEXT    Override the jupyter image. Format: <repo>:<tag>
  -f, --log-file TEXT     Path to log file. If not specified, will log only to
                          stdout
  --mlrun-api-image TEXT  Override the mlrun-api image. Format: <repo>:<tag>
  --mlrun-ui-image TEXT   Override the mlrun-ui image. Format: <repo>:<tag>
  -v, --verbose           Enable debug logging
  --help                  Show this message and exit.
```

### CODE REVIEW

File: automation/common/helpers.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        process.stdin.close()
+
+    if live:
+        for line in iter(process.stdout.readline, b""):
+            stdout += str(line)
+            sys.stdout.write(line.decode(sys.stdout.encoding))
+            if log_file_handler:
+                log_file_handler.write(line.decode(sys.stdout.encoding))
+    else:
+        stdout = process.stdout.read()
```
Comment (TomerShor) [inline automation/common/helpers.py:39]: So if I have a log file handler but `live == False`, the file won't include the logs?

File: automation/deployment/ce.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+@click.option(
+    "--devel",
+    is_flag=True,
+    help="Get the latest RC version of the mlrun chart. (Only works if --chart-version is not specified)",
+)
+@click.option(
+    "-m",
+    "--minikube",
+    is_flag=True,
+    help="Install the mlrun chart in local minikube.",
```
Comment (TomerShor) [inline automation/deployment/ce.py:137]:
```suggestion
    help="Install the mlrun chart in local minikube",
```

File: automation/deployment/ce.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+@click.option(
+    "--set",
+    "set_",
+    help="Set custom values for the mlrun chart. Format: <key>=<value>",
+    multiple=True,
+)
+@click.option(
+    "--upgrade",
+    is_flag=True,
+    help="Upgrade the existing mlrun installation.",
```
Comment (TomerShor) [inline automation/deployment/ce.py:148]:
```suggestion
    help="Upgrade the existing mlrun installation",
```

**Review by TomerShor — CHANGES_REQUESTED** [review]

File: automation/common/helpers.py [hunk 4]
```diff
@@ -29,17 +42,30 @@ def run_command(
...
+    def _write_to_log_file(text: bytes):
+        if log_file_handler:
+            log_file_handler.write(text.decode(sys.stdout.encoding))
+
+    stdout = ""
     if live:
-        for line in iter(process.stdout.readline, b""):
+        for line in iter(stdout_stream.readline, b""):
             stdout += str(line)
             sys.stdout.write(line.decode(sys.stdout.encoding))
```
Comment (TomerShor) [inline automation/common/helpers.py:65]: Will this always write to stdout?
The `log-file` flag is for all logs to be written to the file or **both** the file and stdout?

Comment (quaark) [reply automation/common/helpers.py:65]: I'll clarify in the flag that it's both :)

**Review by Tankilevitch — APPROVED** [review]
Very nice, looks great!

**Pull request merged**